File: Miscellaneous/Other/Review/Stage 2/Review 2/monotonic-stack/monotonic-stack.ts
export function nextGreaterElement(nums: number[]): number[] {
  const result: number[] = new Array(nums.length).fill(-1);
  const monotonicStack: number[] = [];

  for (let i = 0; i < nums.length; i++) {
    while (monotonicStack.length && nums[i] > nums[monotonicStack[monotonicStack.length - 1]]) {
      const index = monotonicStack.pop()!;
      result[index] = nums[i];
    }
    monotonicStack.push(i);
  }

  return result;
}

// test
const nums = [2, 1, 2, 4, 3];
console.log(nextGreaterElement(nums)); // [4, 2, 4, -1, -1]

File: Data Structure & Algorithm/Archive/Template/Monotonic Stack/monotonic-stack.go
package main

import (
	"fmt"
)

func nextGreaterElement(nums []int) []int {
	result := make([]int, len(nums))
	for i := range result {
		result[i] = -1
	}
	monotonicStack := []int{}

	for i := 0; i < len(nums); i++ {
		for len(monotonicStack) > 0 && nums[i] > nums[monotonicStack[len(monotonicStack)-1]] {
			index := monotonicStack[len(monotonicStack)-1]
			monotonicStack = monotonicStack[:len(monotonicStack)-1]
			result[index] = nums[i]
		}
		monotonicStack = append(monotonicStack, i)
	}

	return result
}

func main() {
	nums := []int{2, 1, 2, 4, 3}
	fmt.Println(nextGreaterElement(nums)) // [4, 2, 4, -1, -1]
}